Add collateral return outputs to the API and DB

jonathanknowles opened the pull request:
## Tasks Implemented

ADP-1715 (_Update API to support collateral return outputs_)
ADP-1716 (_Update DB schema to support collateral return outputs_)
ADP-1717 (_Update primitive types to support collateral return outputs_)

## Summary

This PR tackles the "edges" of support for collateral return outputs:
- adjusting API types;
- adjusting primitive types;
- adjusting the DB schema;
- adjusting the DB persistence layer.

The approach used for every commit (except where indicated) in this PR is to:
- first make an adjustment;
- then make the minimal possible number of changes to:
    - get the code to compile;
    - get the tests to pass.
    
This PR does **_not_** update our UTxO transition function to account for collateral outputs. Therefore, when reading blocks with transactions containing collateral outputs that affect a wallet, these outputs will currently be ignored. Updating the UTxO transition function will be handled by ADP-1718.

jonathanknowles committed:
Add `TxCollateralOut` and `TxCollateralOutToken` to database schema.

jonathanknowles committed:
Add field `collateralOutputs` to type `ApiTransaction`.

In the API specification we use an array type for `collateral_outputs`,
for consistency with the encoding of ordinary transaction outputs, and
enabling forward compatibility if in future more than one collateral
output is allowed in a transaction.

But we clearly mark the fact that there can only be zero or one
collateral output. Attempting to parse an `ApiTransaction` with
more than one collateral output will fail with an error.

For the moment, we populate this field with `Nothing`.

jonathanknowles committed:
Format for consistency with team coding standards.

- no trailing whitespace at the end of lines.
- use 4-space indents.
- avoid variable-length indentation.
- limit line lengths to 80 characters where practical.
- use only a single blank line between top-level definitions.

https://input-output-hk.github.io/adrestia/code/Coding-Standards

jonathanknowles committed:
Add field `collateralOutputs` to type `ApiDecodedTransaction`.

In the API specification we use an array type for `collateral_outputs`,
for consistency with the encoding of ordinary transaction outputs, and
enabling forward compatibility if in future more than one collateral
output is allowed in a transaction.

But we clearly mark the fact that there can only be zero or one
collateral output. Attempting to parse an `ApiDecodedTransaction` with
more than one collateral output will fail with an error.

For the moment, we populate this field with `Nothing`.

jonathanknowles committed:
Add field `txInfoCollateralOutput` to type `TransactionInfo`.

In this commit, we simply add the field, and fix up all places that the
compiler requires us to fix.

In many cases, we can simply assign `Nothing` to the field.

In places where we need to assign something other than `Nothing`, we add
a `TODO`.

jonathanknowles committed:
Add field `collateralOutput` to type `Tx`.

Important:

This commit breaks the DB tests, because it provides no way to write or
read collateral outputs to or from the DB. A further commit will provide
proper DB persistence for collateral outputs.

Instead, in this commit, we simply add the field, and fix up all places
that the compiler requires us to fix.

In places where we are providing mock data for tests of transactions
belonging to previous eras, we can simply assign `Nothing` to the field.

In places where we will eventually need to assign something other than
`Nothing`, we add a `TODO`, with a link to an appropriate ticket number.

jonathanknowles committed:
Add utility type `ApiAsArray`.

jonathanknowles committed:
Correct description of `transactionOutputs` in API specification.

sevanspowell reviewed (approved):
Approved with the caveat that someone with more experience takes a look at the database changes. It looks sensible to me but just to be sure.

:+1:

piotr-iohk reviewed (approved):
LGTM.

jonathanknowles committed:
Reduce some of the boilerplate for JSON golden tests.

Instead of using `Proxy`, we can just use type applications.

We also sort the list into alphabetical order to make it easier to spot if
an existing item is already there.

It would be nice if we could generate this list automatically.

jonathanknowles committed:
Add JSON roundtrip test for `ApiAsArray (Maybe t)`.

To keep the size of the golden file small, we pick a type for `t` with
a simple serialization that is unlikely to change.

jonathanknowles committed:
Add JSON parsing failure tests for `ApiAsArray (Maybe t)`.

paolino reviewed (approved):
LGTM

jonathanknowles committed:
Provide DB persistence for collateral outputs.

This commit ensures that all DB tests pass.

jonathanknowles committed:
Rename `collateral` to `collateralInputs`.

We now have two fields relating to collateral within a `Tx`.

To disambiguate them, we now use the following naming scheme everywhere
(where practical):

- inputs
- outputs
- collateral inputs
- collateral outputs

We also change some construction/pattern match sites that use positional
arguments to instead use record syntax. This lowers the risk that fields
with identical types are accidentally juxtaposed.

jonathanknowles committed:
Add field `collateralOutputs` to `Buildable` instance for type `Tx`.

jonathanknowles committed:
Add field `txCollateralOutput` to type `MkApiTransactionParams`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -4124,3 +4135,28 @@ instance FromJSON (ApiT TxScriptValidity) where
+        [] ->
+            pure $ ApiAsArray Nothing
+        _  ->
+            fail "Expected at most one item."
```
Suggestion: Improve this error message to make it more useful for the user.

It could be more useful if we were able to report something like:

> Expected at most one item for "collateral_outputs".

jonathanknowles replied:
Fixed in 3cd2c4e99cf9ba9b228ae0b2d344d63cbd7a2aa6.

jonathanknowles committed:
Make error message more useful on failure to parse `ApiAsArray`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3251/files#r859371410

jonathanknowles committed:
Factor out common type expression into `ApiTxOutput` type synonym.
